### PR TITLE
Refactor dependency calculations

### LINKED ISSUE #3: Support projects with empty root
If project has no Go files inside root directory `manul -Q` will exit with the following message:

```
no buildable Go source files in /path/to/repo
```

For go tool it is solved with `./...` argument which means walk current directory recursively.

### LINKED ISSUE #15: Dependencies in the test files are ignored
The dependencies in the test files are ignored when searching for imports. In the imports.go only Package.Imports are being considered. I guess even Package.TestImports should be included. 

Thanks,
Koushik Shetty

### LINKED ISSUE #26: not working on mac osx
test file 

```
package main

import "github.com/kovetskiy/manul-test-foo"

func main() {
    foo.A()
}
```

run command  throw error, but "all dependencies already vendored" in the end.

```
 $ /usr/local/bin/manul -I
2016/10/23 00:41:47 cannot find package "github.com/kovetskiy/manul-test-foo" in any of:
    /usr/local/Cellar/go/1.7.3/libexec/src/github.com/kovetskiy/manul-test-foo (from $GOROOT)
    /opt/dev/fxquant/go/src/github.com/kovetskiy/manul-test-foo (from $GOPATH)
all dependencies already vendored
```

version

```
$ /usr/local/bin/manul --version
manul 1.4
```

### LINKED ISSUE #27: manul ignores build tags?
This is on mac.

You can repro with this code: https://github.com/kjk/web-blog (`git clone --recursive https://github.com/kjk/web-blog`)

```
$ manul -U -r
2016/10/22 14:30:37 updating dependency github.com/garyburd/go-oauth
2016/10/22 14:30:37 updating dependency github.com/go-fsnotify/fsnotify
2016/10/22 14:30:38 updating dependency github.com/gorilla/securecookie
2016/10/22 14:30:38 updating dependency github.com/gorilla/websocket
2016/10/22 14:30:39 updating dependency github.com/kjk/u
2016/10/22 14:30:39 updating dependency github.com/kr/fs
2016/10/22 14:30:40 updating dependency github.com/microcosm-cc/bluemonday
2016/10/22 14:30:40 updating dependency github.com/russross/blackfriday
2016/10/22 14:30:41 updating dependency github.com/shurcooL/sanitized_anchor_name
2016/10/22 14:30:42 updating dependency github.com/thomas11/atomgenerator
2016/10/22 14:30:42 updating dependency golang.org/x/net
updated 11 dependencies
$ go build
vendor/github.com/go-fsnotify/fsnotify/kqueue.go:18:2: cannot find package "golang.org/x/sys/unix" in any of:
    /Users/kjk/src/go/src/github.com/kjk/blog/vendor/golang.org/x/sys/unix (vendor tree)
    /usr/local/Cellar/go/1.7.3/libexec/src/golang.org/x/sys/unix (from $GOROOT)
    /Users/kjk/src/go/src/golang.org/x/sys/unix (from $GOPATH)
kjkmacpro:blog kjk$
```

The problem seems to be that in `kqueue.go` we have:

```
// +build freebsd openbsd netbsd dragonfly darwin

package fsnotify

import (
    "errors"
    "fmt"
    "io/ioutil"
    "os"
    "path/filepath"
    "sync"
    "time"

    "golang.org/x/sys/unix"
)
```

I assume it's the build tag. Since I'm building on mac and build tag specifies darwin, the file should be parsed and `golang.org/x/sys/unix` be vendored as well.

### LINKED ISSUE #30: Recursive dependencies with own vendoring
Given the following example:

```
# start within a new, empty $GOPATH

# install manul
go get github.com/kovetskiy/manul

# create an example project importing fsouza/go-dockerclient
mkdir -p src/github.com/example/example
cd src/github.com/example/example
cat > main.go <<EOF
package main
import "github.com/fsouza/go-dockerclient"
EOF

# get all dependencies recursively
go get ./...

# query manul
manul -Q
```

Which outputs:

```
2017/01/31 22:31:53 cannot find package "github.com/docker/go-connections/nat" in any of:
	/usr/local/Cellar/go/1.7.5/libexec/src/github.com/docker/go-connections/nat (from $GOROOT)
	<my-gopath>/src/github.com/docker/go-connections/nat (from $GOPATH)
github.com/docker/docker
github.com/fsouza/go-dockerclient
```

First of all, why is `github.com/docker/docker` even there? It is a dependency of a dependency, but I did not add `-r` to my command for recursive querying. Does manul do this by default? If so, what's the use of the `-r` flag?

So the recursive dependency process is stopped at the first package it could not find. It complains about `github.com/docker/go-connections/nat`, but in fact there are more.

The reason `github.com/docker/go-connections/nat` was not installed by the `go get` command, is because this dependency is already vendored in the repo that contains the package that imports it (`github.com/docker/docker`).

Manul, however, wants the package to be installed and wants to vendor it. Of course, I can manually go get `github.com/docker/go-connections/nat` (and the others that follow afterwards) and subsequently vendor them with `manul -I`, but then I would end up with an unnecessary, unused vendor package that doesn't even match with the intended version (`<...>/docker/docker/vendor/` might contain a different version).

Am I correct in thinking that the solution for this would be to have manul detect dependencies that are already vendored by the dependency that imports it, just like `go get` does when doing recursive installs?

### PR DESCRIPTION
**Pull request by johansglock**
- Now honors recursive flag (fixes kovetskiy/manul#30)
- Option to include testing dependencies (fixes kovetskiy/manul#15)
- Don't include recursive deps if they are already vendored (fixes kovetskiy/manul#3
- Don't vendor unused packages (fixes kovetskiy/manul#30)
- Do a got get at the beginning as a usability feature (fixes kovetskiy/manul#26)
- Should also fix kovetskiy/manul#27, though unconfirmed

### CODE REVIEW

**Comment by johansglock** [comment]
FYI, the tests fail in travis, I ran them manually locally, all output is what it's expected to be, though the test script itself fails.

File: imports.go [hunk 1]
```diff
@@ -1,135 +1,167 @@
...
-		_, ok := imports[importing]
-		if !ok {
-			err = recursiveParseImports(imports, importing, cwd)
+	testImports = unique(testImports)
+	if recursive {
+		for _, i := range testImports {
+			testData, err := list(i)
 			if err != nil {
-				return err
+				return imports, nil
```
Comment (kovetskiy) [inline imports.go]: why error is not returns here?

File: imports.go [hunk 2]
```diff
@@ -1,135 +1,167 @@
...
 }
 
-func parseImports(recursive bool) ([]string, error) {
-	cwd, err := os.Getwd()
-	if err != nil {
-		return nil, err
-	}
+func filterPackages(packages []string) []string {
+	var imports []string
+	cwd, _ := os.Getwd()
```
Comment (kovetskiy) [inline imports.go]: please do not ignore errors, better to return error than ignore it and produce abnormal output

File: imports.go [hunk 3]
```diff
@@ -1,135 +1,167 @@
...
+		args = append(args, "-t")
+	}
+	args = append(args, packages...)
 
-			if inTests {
-				importpath = strings.Replace(importpath, "__blankd__", "localhost:60001", -1)
-			}
+	out, err := execute(exec.Command("go", args...))
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%s\n", out)
```
Comment (kovetskiy) [inline imports.go]: do not print on such nested function level, return error and print it where function has called

File: imports.go [hunk 4]
```diff
@@ -1,135 +1,167 @@
...
+	if err != nil {
+		return data, err
+	}
 
-			if found {
-				continue
-			}
+	err = json.Unmarshal([]byte(out), &data)
+	if err != nil {
+		return data, err
```
Comment (kovetskiy) [inline imports.go]: json.Unmarshal returns abmigious error when can't unmarshal json payload, so, please, add value from `out` variable to error message, like as following:
```
return data, fmt.Errof("can't unmarshal go list JSON output: %s\n%q", err, out)
```

File: imports.go [hunk 5]
```diff
@@ -140,3 +172,23 @@ func isOwnPackage(path, cwd string) bool {
 	}
 	return false
 }
+
+func unique(input []string) []string {
+	u := make([]string, 0, len(input))
+	m := make(map[string]bool)
```
Comment (kovetskiy) [inline imports.go]: please do not use one-symbol variables, it is more clear when name of variable describes what the purpose (or meaning) of variable. 
u -> result
m -> hash or table

File: main.go [hunk 6]
```diff
@@ -40,6 +40,7 @@ Options:
     -Q --query      List all dependencies.
         -o          List only already-vendored dependencies.
     -C --clean      Detect all unused vendored dependencies and remove it.
+    -t --testing    Include testing dependencies.
```
Comment (kovetskiy) [inline main.go]: may be here we should write something like `Include dependencies from tests?` What you think about it?

File: submodule.go [hunk 7]
```diff
@@ -190,6 +190,9 @@ func getRootImportpath(importpath string) (string, error) {
 		return "", err
 	}
 
+	cwd, _ := os.Getwd()
```
Comment (kovetskiy) [inline submodule.go]: do not ignore error here.

Actually, it is 3th call of os.Getwd() may be create global variable `workdir` and initialize it in `init()` function?

File: imports.go [hunk 8]
```diff
@@ -1,135 +1,167 @@
...
+
+}
+
+func calculateDependencies(packages []string, recursive, testDependencies bool) ([]string, error) {
+	var deps, imports, testImports, testDeps []string
+
+	for _, pkg := range packages {
+		data, err := list(pkg)
+		if err != nil {
+			return imports, nil
```
Comment (kovetskiy) [inline imports.go]: why error is ignored?

File: imports.go (hunk 1 above)
Comment (kovetskiy) [reply imports.go]: and please wrap error with some useful message like (`error listing import: %s`, importName)

File: imports.go [hunk 9]
```diff
@@ -1,135 +1,167 @@
...
 	}
 
-	if path != "." {
-		standard := false
+	imports = filterPackages(imports)
+
+	sort.Strings(imports)
+
+	return imports, nil
+
```
Comment (kovetskiy) [inline imports.go]: please remove extra new line after last return statement

File: imports.go [hunk 10]
```diff
@@ -140,3 +172,23 @@ func isOwnPackage(path, cwd string) bool {
...
+		if _, ok := m[val]; !ok {
+			m[val] = true
+			u = append(u, val)
+		}
+	}
+
+	return u
+}
+
+type listOutput struct {
```
Comment (kovetskiy) [inline imports.go]: move this declaration to begin of the file, it is more obvious to find declaration of such structures when they are defined in the beginning of file

**Review by kovetskiy — CHANGES_REQUESTED** [review]

**Comment by kovetskiy** [comment]
Hello. Thank you for the pull-request. I'm glad to see contributions like this.

Can you please add some testcases (for example for #30)?

**Comment by johansglock** [comment]
Oh, the error handling was probably a last minute copy-paste mistake :) I will get around to fixing all of the requested changed later.

Regarding the tests; I would love to, if I could get the testing setup to run to begin with. Travis is failing as well. Any insights on these?

**Comment by kovetskiy** [comment]
@johansglock I've fixed running tests in travis, now everything works as expected: https://travis-ci.org/kovetskiy/manul/builds/202185390

File: imports.go [hunk 11]
```diff
@@ -11,14 +11,25 @@ import (
...
 	if err != nil {
 		return imports, fmt.Errorf("error listing packages: %s", err)
 	}
 
-	ensureDependenciesExist(packages, testDependencies)
+	// Ensuring our dependencies exists isn't a strict requirement, therefore
+	// only print a message to stderr rather then completely failing
+	err = ensureDependenciesExist(packages, testDependencies)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%s", err)
```
Comment (kovetskiy) [inline imports.go]: here should also return error on top level and print it in place where function parseImports has called

Comment (johansglock) [reply imports.go]: I think I'll just omit the print, there is not really a distinction between warnings and errors, so let's just omit the warning?

What do you think?

Still processing the rest btw.

Comment (kovetskiy) [reply imports.go]: Hmm, okay, keep it as is. anyway, you should replace calling from `Printf(stderr, "%s", err)` to `Println(err)` or add `\n` after `%s` for adding new line after error message.

Comment (johansglock) [reply imports.go]: Error is a propagation from the go commandline utility which already includes a newline.

Comment (johansglock) [reply imports.go]: Though I included it before; I'll add it again, it's nice to have the extra newline between the error and regular output

File: main.go [hunk 12]
```diff
@@ -69,6 +70,12 @@ func init() {
 		}
 	}
 	os.Args = newArgs
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		panic(fmt.Errorf("couldn't get current working directory: %s", err))
```
Comment (kovetskiy) [inline main.go]: I guess there is no need in panic's stack trace, just print error to os.Stderr and os.Exit(1)

Comment (johansglock) [reply main.go]: and os.Exit? this error is sort of crucial

Comment (kovetskiy) [reply main.go]: yes

File: imports.go [hunk 13]
```diff
@@ -1,142 +1,199 @@
...
-		_, ok := imports[importing]
-		if !ok {
-			err = recursiveParseImports(imports, importing, cwd)
+	testImports = unique(testImports)
+	if recursive {
+		for _, testImport := range testImports {
+			testData, err := list(testImport)
 			if err != nil {
-				return err
+				return imports, fmt.Errorf("failed to list dependencies for testing package %s: %s", testImport, err)
```
Comment (kovetskiy) [inline imports.go]: please format line with max width as 80 symbols like as in existing code

File: imports.go [hunk 14]
```diff
@@ -1,142 +1,199 @@
...
+		args = append(args, "-t")
+	}
+	args = append(args, packages...)
 
-			if inTests {
-				importpath = strings.Replace(importpath, "__blankd__", "localhost:60001", -1)
-			}
+	out, err := execute(exec.Command("go", args...))
+	if err != nil {
+		return fmt.Errorf("couldn't 'go get' dependencies:\n%s", out)
```
Comment (kovetskiy) [inline imports.go]: on 143 line you use `can't` may be here replace `could` to `can`?

File: imports.go [hunk 15]
```diff
@@ -1,142 +1,199 @@
...
+func listPackages() ([]string, error) {
+	var packages []string
+
+	out, err := execute(exec.Command("go", "list", "-e", "./..."))
+	if err != nil {
+		return packages, err
+	}
+
+	r := strings.NewReader(out)
+	s := bufio.NewScanner(r)
```
Comment (kovetskiy) [inline imports.go]: wow, reader and scanner, may be just use strings.Split(strings.TrimSpaces(out), "\n")?

**Comment by johansglock** [comment]
@robbertkl can you come up with some nice tests?

**Comment by robbertkl** [comment]
@johansglock Who are you? Do I know you?

Not sure why I'm mentioned in this PR....

**Comment by kovetskiy** [comment]
@johansglock I can add tests if you haven't a time to do it, can I merge it now?

**Comment by johansglock** [comment]
This is safe to be merged yeah; haven't had time for the tests. So if you would be willing to do that, that would be great yeah!

**Review by kovetskiy — APPROVED** [review]

**Comment by kovetskiy** [comment]
Merged, thanks for the contribution!

**Pull request merged**